### PR TITLE
Add Lprologue

### PR DESCRIPTION
**Pull request by mshinwell**
The forthcoming gdb support needs to be able to insert items in the linearised instruction stream before the function prologue.  This patch, which consists almost entirely of code moves and should be a no-op, adds a new `Lprologue` instruction which causes emission of the prologue.  This enables other instructions (which will only be used for debugging information emission) to be placed before it.  The size calculations for the prologue on ARM, AArch64 and POWER platforms are probably not strictly required but have been included for completeness.

### CODE REVIEW

File: asmcomp/arm64/emit.mlp [hunk 1]
```diff
@@ -419,11 +419,16 @@ module BR = Branch_relaxation.Make (struct
 
   let offset_pc_at_branch = 0
 
+  let prologue_size () =
+    (if frame_size () > 0 then 2 else 0)
```
Comment (chambart) [inline asmcomp/arm64/emit.mlp:423]: This is a slight over-approximation, but this shouldn't matter

File: asmcomp/amd64/emit.mlp [hunk 2]
```diff
@@ -911,21 +926,6 @@ let fundecl fundecl =
   D.label (emit_symbol fundecl.fun_name);
   emit_debug_info fundecl.fun_dbg;
   cfi_startproc ();
```
Comment (chambart) [inline asmcomp/amd64/emit.mlp:928]: Wouldn't it make sense for `cfi_startproc` to be part of what is emitted by `Lprelude`. Of course this depends on what kind of things you consider adding before the prelude.

**Review by chambart — APPROVED** [review]
This patch should effectively be a noop.

What do you consider adding before the prologue ? Isn't that better handled by adding some more field to fundecl ?

**Comment by mshinwell** [comment]
As far as I recall the reason for needing this relates to the fact that the DWARF available range information is emitted based on labels in the linearised code (i.e. values of type `Linearize.t`).  The address of the earliest possible such label will be after the prologue if you don't have this change.  The consequence is that when you arrive at the top of a function in the debugger, you won't see your parameters...

File: asmcomp/amd64/emit.mlp (hunk 2 above)
Comment (mshinwell) [reply asmcomp/amd64/emit.mlp:928]: I will check this.

File: asmcomp/arm64/emit.mlp (hunk 1 above)
Comment (mshinwell) [reply asmcomp/arm64/emit.mlp:423]: Indeed -- I followed the existing size calculation code though, which uses this value.

File: asmcomp/amd64/emit.mlp (hunk 2 above)
Comment (mshinwell) [reply asmcomp/amd64/emit.mlp:928]: I think this is ok as is, having thought about it this morning.  It matches up nicely with `cfi_endproc` at the moment; and it also seems in some sense correct that it's right at the top in the assembly output.

**Comment by mshinwell** [comment]
@chambart has approved and all comments addressed without needing code changes, so merging.

**Pull request merged**